Add new RedOS Linux

foxzi opened the pull request:
Add new RedOS Linux https://redos.red-soft.ru/. This is linux rpm based distro.
 Createng RedOS image uses multistage building based on CentOS 7.

foxzi committed:
Add RedOS 7.3

It is RPM based Linux distribution.

The RedOS image is built using the multistage build process. CentOS 7 is
used as the builder container: it calls `yum --installroot=<...>
install` to install packages into the root of the target RedOS image
(the idea is similar to chroot).

This way we're able to create RedOS image without extra scripting around
`docker build` command. It seems, it is not possible if we'll use the
official RedOS docker image archive (because it is not in a docker
registry, just a file).

https://redos.red-soft.ru

Co-authored-by: Alexander Turenko <alexander.turenko@tarantool.org>

Totktonada commented on redos/7/Dockerfile:
```diff
@@ -0,0 +1,57 @@
+RUN yum -y --installroot=$CHROOT_DIR install -y \
+    yum \
+    rpm-devel \
+    rpm-devel \
```
Nit: `rpm-devel` is selected twice.

NickVolynkin commented on redos/7/Dockerfile:
```diff
@@ -0,0 +1,57 @@
+    python3-zope-interface \
+    python3-greenlet
+
+RUN sed -i.bak -n -e '/^Defaults.*requiretty/ { s/^/# /;};/^%wheel.*ALL$/ { s/^/# / ;} ;/^#.*wheel.*NOPASSWD/ { s/^#[ ]*//;};p' $CHROOT_DIR/etc/sudoers
```
`-i.bak` — how will we use the backup in the image?

Totktonada replied:
Left just `-i`.

NickVolynkin commented on redos/7/Dockerfile:
```diff
@@ -0,0 +1,57 @@
+COPY --from=prepare /chroot/. /
+
+#Rpm packages in Redos repos are broken, so we need to install them manually
+RUN pip3 install gevent
```
```suggestion
RUN pip3 install gevent --no-cache-dir
```

foxzi replied:
done

NickVolynkin commented on redos/7/Dockerfile:
```diff
@@ -0,0 +1,57 @@
+    python3-pip \
+    python3-devel \
+    python3-zope-interface \
+    python3-greenlet
```
why don't we remove yum caches after installing?

NickVolynkin replied:
Oh, it's done later, I see

Totktonada commented on redos/7/Dockerfile:
```diff
@@ -0,0 +1,57 @@
+    rpm-devel \
+    rpm-devel \
+    devtoolset \
+	rpmdevtools \
```
Nit: Tab/space mix here and around gettext-devel below.

foxzi replied:
fixed

Totktonada commented on redos/7/Dockerfile:
```diff
@@ -0,0 +1,57 @@
+    git \
+    sudo \
+    yum-utils \
+    zypper \
```
`zypper` is the package manager openSUSE, it is a bit surprising to see it here.

foxzi replied:
I remove this package

Totktonada commented on redos/7/Dockerfile:
```diff
@@ -0,0 +1,57 @@
+    python3-zope-interface \
+    python3-greenlet
+
+RUN sed -i.bak -n -e '/^Defaults.*requiretty/ { s/^/# /;};/^%wheel.*ALL$/ { s/^/# / ;} ;/^#.*wheel.*NOPASSWD/ { s/^#[ ]*//;};p' $CHROOT_DIR/etc/sudoers
```
Nit: The command itself is not very obvious, I would leave a comment as in other docker files (at least in centos/7, centos/8).

foxzi replied:
done

Totktonada commented on redos/7/Dockerfile:
```diff
@@ -0,0 +1,57 @@
+FROM centos:7 as prepare
```
How about ABI compatibility between RedOS 7.2 and RedOS 7.3? Shouldn't we place the docker file into `redos/7.3` instead of `redos/7` and tag the image as `packpack/packpack:redos-7.3`?

I don't know, just asking.

Totktonada replied:
It seems, it is resolved. Now the directory is named as `redos/7.3`.

Totktonada commented on redos/7/Dockerfile:
```diff
@@ -0,0 +1,57 @@
+    yum \
+    rpm-devel \
+    rpm-devel \
+    devtoolset \
```
devtoolset is installed, but not set up? We install and setup devtoolset on CentOS 7 to use GCC 8 instead of default 4.8.5:

https://github.com/packpack/packpack-docker-images/blob/948e9f05366acac80d9d0102c3c5a8dd676240af/centos/7/Dockerfile#L42-L49

But RedOS 7.3 mirror provides GCC 8 on its own.

Totktonada replied:
Removed devtoolset.

Totktonada commented on redos/7/Dockerfile:
```diff
@@ -0,0 +1,57 @@
+
+RUN yum -y --installroot=$CHROOT_DIR update
+
+RUN yum -y --installroot=$CHROOT_DIR install -y \
```
How packages are chosen? Can we do `yum -y groupinstall 'Development Tools'` similarly to CentOS 7 image?

I think a short comment about motivation behind this package list would be helpful.

If the motivation is to install everything that is needed for building tarantool, running tests and create an RPM package, it lacks zlib-devel and python3-pyyaml at least.

Totktonada replied:
> <...> it lacks zlib-devel and python3-pyyaml at least.

If I run `make -f ~/packpack/pack/Makefile -C . BUILDDIR=~/build -j` (see https://github.com/tarantool/infra/issues/106) from inside the container and several more packages are installed: libunwind-devel (not a real dependency and I guess it'll be removed in https://github.com/tarantool/tarantool/issues/6998), perl-Test-Harness and python3-gevent.

It is okay per se. I just want to understand a rule how this list was created.

foxzi replied:
Group packages Development Tools has 2 subgroups mandatory, default and optional. If I install all of them that docker images increases size to 1,7Gb from 0.8Gb. I checked witch packages needs for building tarantool and install them separately.

Totktonada replied:
I splitted general development packages from packages necessary to build gevent using pip. Added explanations as comments. Now it should be clear.

Totktonada commented on redos/7/Dockerfile:
```diff
@@ -0,0 +1,57 @@
+
+RUN yum -y --installroot=$CHROOT_DIR update
+
+RUN yum -y --installroot=$CHROOT_DIR install -y \
```
I guess that some of the packages are from CentOS 7 itself (at least I see no `zypper` on the Yandex's RedOS mirror). Or at least they can be installed here. Is it deliberately? If so, please, add a comment, which will answer the following questions:

* Is it okay to place packages from CentOS 7 here (not from RedOS)?
* If so, how we should limit it to don't end up with CentOS 7 under RedOS name?
* If there are any other caveats with packages selection.

If only RedOS 7.3 packages should be allowed here, we should let the code fail if a package not from RedOS 7.3 is selected. I guess we can remove all mirrors except RedOS's one. If it is not so, I would like to see an explanation regarding it either in the code or in the commit message.

----

Update: I run the build and see the following warnings:

```
No package devtoolset available.
Package yum-utils is obsoleted by dnf-utils, trying to install dnf-utils-4.0.18-3.el7.noarch instead
No package zypper available.
```

So I guess the following:

* Packages from CentOS 7 are not installed here (and I don't know how it works, TBH; `--installroot` uses only chroot's mirrors?)
* Non-existing selections are ignored (it would be good to fail in this case).
* `zypper` and `devtoolset` actually not installed.
* It worth to install dnf-utils instead of yum-utils.

Despite that everything looks mostly good I would like to see some explanations (I kept my original message above, so you can see what was confusing for me).

foxzi replied:
I replaced yum-utils with dnf-utils and removed devtoolset.

Totktonada replied:
It seems, it was not done, but I did.

Totktonada commented on redos/7/Dockerfile:
```diff
@@ -0,0 +1,57 @@
+
+COPY --from=prepare /chroot/. /
+
+#Rpm packages in Redos repos are broken, so we need to install them manually
```
All packages or just python3-gevent?

foxzi replied:
Package python3-gevent in redos is broken. It is easiest way to set it. But there are 3 available points how to fix this problem:
 *write to support redos
 *create backport repo with self made package
 *use pip

Totktonada replied:
I clarified this comment.

Totktonada commented on redos/7/Dockerfile:
```diff
@@ -0,0 +1,57 @@
+COPY --from=prepare /chroot/. /
+
+#Rpm packages in Redos repos are broken, so we need to install them manually
+RUN pip3 install gevent
```
It would be convenient to have `CMD ["/bin/bash"]` at the end to run a container without explicit `/bin/bash`.

Totktonada reviewed (commented):
Thanks for the patch!

My review may look picky, but mostly I just want to understand and hold details: so I'm asking questions and requesting for explanations.

ylobankov commented on redos/7.3/Dockerfile:
```diff
@@ -0,0 +1,58 @@
+
+RUN mkdir -p $CHROOT_DIR
+
+#Download any available meta package redos-release
```
Please add a space between `#` and `Download`.

Totktonada replied:
Done.

ylobankov commented on redos/7.3/Dockerfile:
```diff
@@ -0,0 +1,58 @@
+
+#Download any available meta package redos-release
+RUN for rpm in `curl -s $REPO_URL | grep 'redos-release' | sed 's/<.*>\(.*\)<\/.*>/\1/' | awk '{print $1}'`; \
+    do \
```
Usually `do` goes on one level with `for`:

```
for i in $(ls /tmp/); do 
    ...
done
```

Totktonada replied:
I started to follow this recommendation, but ended up with some rewritting of this block. A think of taste, but lines now shorter. I have added myself into co-authors to let people blame me about my mistakes.

ylobankov commented on redos/7.3/Dockerfile:
```diff
@@ -0,0 +1,58 @@
+RUN for rpm in `curl -s $REPO_URL | grep 'redos-release' | sed 's/<.*>\(.*\)<\/.*>/\1/' | awk '{print $1}'`; \
+    do \
+        if [[ $rpm =~ redos-release-[0-9].[0-9]-([0-9]|[0-9][0-9]).el7.noarch.rpm ]]; \
+        then \
```
The same as for `do`.

Totktonada replied:
Rewritten (see comment above).

ylobankov commented on redos/7.3/Dockerfile:
```diff
@@ -0,0 +1,58 @@
+        fi; \
+    done
+
+#setup base system
```
Please keep consistency around the comments: `s` -> `S`, add a space.

Totktonada replied:
I have splitted this block and expended comments.

ylobankov commented on redos/7.3/Dockerfile:
```diff
@@ -0,0 +1,58 @@
+# Enable sudo without tty
+RUN sed -i.bak -n -e '/^Defaults.*requiretty/ { s/^/# /;};/^%wheel.*ALL$/ { s/^/# / ;} ;/^#.*wheel.*NOPASSWD/ { s/^#[ ]*//;};p' $CHROOT_DIR/etc/sudoers
+
+#Clear cache
```
Add a space.

Totktonada replied:
Done.

ylobankov commented on redos/7.3/Dockerfile:
```diff
@@ -0,0 +1,58 @@
+
+COPY --from=prepare /chroot/. /
+
+#Rpm packages in Redos repos are broken, so we need to install them manually
```
Add a space.

Totktonada replied:
Done (I rewrote the comment).

Totktonada reviewed (approved):
I'm a co-author now (made changes myself), so my review is not sufficient. @ylobankov, can you, please, take another look on the patch?

ylobankov reviewed (approved):
LGTM

Totktonada commented:
Yaroslav said that he is working on building python3-gevent package and enabling a repository with it. So it is on Yaroslav now.

ylobankov committed:
Use backports repo for python3-gevent in RedOS 7.3

The `python3-gevent` package was build from the source RPM file
(version 21.1.2) and uploaded to the backports repo. So there is
no need to install this package via `pip` now, it will be installed
while the build process from the backports repo.

Totktonada commented:
The updated version of the patch seems okay too. Thank you all for this work!
